File: app/schemas/booking.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, model_validator

from app.models.booking import BookingStatus, BookingType
from app.schemas.recommendation import (
    DecisionReport,
    RecommendationFilters,
    RecommendationPreferences,
    RecommendationWeights,
)


class BookingCreate(BaseModel):
    start_time: datetime
    end_time: datetime
    parking_spot_id: int | None = None
    type: BookingType = BookingType.guest
    vehicle_id: int | None = None
    plate_number: str | None = None

    # Auto-assign mode
    auto_assign: bool = False
    parking_lot_id: int | None = None
    recommendation_filters: RecommendationFilters | None = None
    recommendation_preferences: RecommendationPreferences | None = None
    recommendation_weights: RecommendationWeights | None = None

    @model_validator(mode="after")
    def validate_assignment_mode(self):
        if self.auto_assign:
            if self.parking_spot_id is not None:
                raise ValueError("Provide either parking_spot_id or auto_assign=true, not both")
            if self.parking_lot_id is None:
                raise ValueError("parking_lot_id is required when auto_assign=true")
        elif self.parking_spot_id is None:
            raise ValueError("parking_spot_id is required when auto_assign=false")

        return self

    class Config:
        from_attributes = True


class BookingUpdate(BaseModel):
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    type: Optional[BookingType] = None
    status: Optional[BookingStatus] = None


class BookingOut(BaseModel):
    id: int
    user_id: int
    parking_spot_id: int
    vehicle_id: int | None = None
    type: BookingType
    status: BookingStatus
    plate_number: str | None = None
    start_time: datetime
    end_time: datetime
    assignment_mode: str = "manual"
    assignment_explanation: str | None = None
    assignment_metadata: dict | None = None
    decision_report: DecisionReport | None = None

    class Config:
        from_attributes = True

File: frontend/src/shared/types/booking.ts
import type { BookingStatus, ListResponse, SortOrder, SpotType, VehicleType } from './common';
import type { DecisionReport } from './recommendation';

export type BookingType = 'guest' | 'rental';

export interface Booking {
  id: number;
  user_id: number;
  parking_spot_id: number;
  vehicle_id?: number | null;
  type: BookingType;
  status: BookingStatus;
  plate_number?: string | null;
  start_time: string;
  end_time: string;
  assignment_mode: 'manual' | 'auto' | string;
  assignment_explanation: string | null;
  assignment_metadata?: Record<string, unknown> | null;
  decision_report?: DecisionReport | null;
}

export interface RecommendationWeights {
  availability?: number;
  spot_type?: number;
  zone?: number;
  charger?: number;
  role?: number;
  conflict?: number;
}

export interface RecommendationFilters {
  spot_types?: SpotType[];
  zone_ids?: number[];
  vehicle_type?: VehicleType;
  size_category?: 'small' | 'medium' | 'large';
  requires_charger?: boolean;
}

export interface RecommendationPreferences {
  preferred_spot_types?: SpotType[];
  preferred_zone_ids?: number[];
  prefer_charger?: boolean;
  needs_accessible_spot?: boolean;
  max_results?: number;
}

export interface CreateBookingPayload {
  start_time: string;
  end_time: string;
  parking_spot_id?: number;
  vehicle_id?: number;
  type?: BookingType;
  plate_number?: string;
  auto_assign?: boolean;
  parking_lot_id?: number;
  recommendation_filters?: RecommendationFilters;
  recommendation_preferences?: RecommendationPreferences;
  recommendation_weights?: RecommendationWeights;
}

export interface UpdateBookingPayload {
  start_time?: string;
  end_time?: string;
  type?: BookingType;
  plate_number?: string;
  status?: BookingStatus;
}

export interface BookingsQuery {
  mine?: boolean;
  parking_lot_id?: number;
  parking_spot_id?: number;
  vehicle_id?: number;
  from?: string;
  to?: string;
  status?: BookingStatus;
  statuses?: BookingStatus[];
  limit?: number;
  offset?: number;
  sort_by?: 'start_time' | 'end_time' | 'status' | 'id';
  sort_order?: SortOrder;
}

export type BookingListResponse = ListResponse<Booking>;
